syntax = "proto3";

message XTableMessage {
  Command command = 1;
  optional string key = 2;
  optional bytes value = 3;
  optional Type type = 4;
  optional bytes id = 5;

  enum Type {
    UNKNOWN = 0;
    STRING = 1;
    DOUBLE = 2;
    INT64 = 5;
    BOOL = 6;
    BYTES = 7;
    ENUM = 8;
    MESSAGE = 9;
    ARRAY = 10;
    OBJECT = 11;
  }
  enum Command {
    UNKNOWN_COMMAND = 0;
    PUT = 1;
    GET = 2;
    DEBUG = 3;
    GET_TABLES = 4;
    RUN_SCRIPT = 5;
    UPDATE_KEY = 6;
    DELETE = 7;
    PUBLISH = 8;
    SUBSCRIBE_DELETE = 9;
    UNSUBSCRIBE_DELETE = 10;
    UNSUBSCRIBE_UPDATE = 11;
    PING = 12;
    GET_RAW_JSON = 13;
    DELETE_EVENT = 14;
    UPDATE_EVENT = 15;
    INFORMATION = 16;
    REBOOT_SERVER = 17;
    REGISTRY = 18;
  }
  message XTableLog {
    Level level = 1;
    string message = 2;
    enum Level {
      UNKNOWN = 0;
      INFO = 1;
      WARNING = 2;
      SEVERE = 3;
      FATAL = 4;
    }
  }
  message XTableUpdate {
    string key = 1;
    Category category = 2;
    bytes value = 3;
    XTableMessage.Type type = 4;
    enum Category {
      UNKNOWN = 0;
      UPDATE = 1;
      DELETE = 2;
      PUBLISH = 3;
      REGISTRY = 4;
      INFORMATION = 5;
      LOG = 6;
    }
  }

  message ClientInfo {
    string ip_address = 1;
    string hostname = 2;
    string operating_system = 3;
    int32 port = 4;
    bool is_active = 5;
    string version = 6;
    string user_agent = 7;
  }
}
